Cython/Compiler/ExprNodes.py

```diff
@@ -13878,6 +13892,52 @@ def analyse_type_annotation(self, env, assigned_value=None):
             warning(annotation.pos, "Unknown type declaration in annotation, ignoring")
         return base_type, arg_type
 
+class AssignmentExpressionNode(ExprNode):
+    """
+    Also known as a named expression or the walrus operator
+
+    Arguments
+    lhs - NameNode
+    rhs - ExprNode  (not stored as an attribute because it is likely to be heavily transformed)
+
+    internally uses
+    impl - UtilNodes.TempResultFromStatNode
+    """
+    # unusually, child_attrs and subexprs are deliberately not the same because it's
+    # sort of a hybrid node
+    subexprs = ["impl"]
+
+    impl = None
+    is_temp = True
+
+    def __init__(self, pos, lhs, rhs, **kwds):
+        from .UtilNodes import TempResultFromStatNode, ResultRefWithTypePropertyNode
+        super(AssignmentExpressionNode, self).__init__(pos, lhs=lhs, **kwds)
+
+        temp = ResultRefWithTypePropertyNode(pos=pos,
+                    type_getter = lambda: self.impl.body.stats[0].rhs.type if self.impl else None)
+        assignment1 = Nodes.SingleAssignmentNode(pos=pos, lhs=temp, rhs=rhs, first=True)
+        assignment2 = Nodes.SingleAssignmentNode(pos=pos, lhs=lhs, rhs=temp, first=True,
+                                                 is_assignment_expression=True)
```

Comment:
Two assignment nodes? It feels like we should try with just one. Maybe the `AssignmentExpressionNode` needs to do more work then, and probably also treated differently in the control flow analyser etc., but it should be possible to pass a single temp value through as the `result_code` *and* do the assignment. You may have to play a bit with the temp disposal, but I faintly recall having done that before in some other node, probably some utility node.